File: server/controllers/order_controller.py
from fastapi import HTTPException
from pymongo import MongoClient
from bson import ObjectId
from datetime import datetime
import os

client = MongoClient(os.getenv("DATABASE_URL", "mongodb://localhost:27017"))
db = client[os.getenv("MONGO_DB_NAME", "ecommerce")]

def _serialize(doc: dict):
    if doc is None:
        return None
    def _walk(o):
        if isinstance(o, ObjectId):
            return str(o)
        if isinstance(o, datetime):
            return o.isoformat()
        if isinstance(o, dict):
            return {k: _walk(v) for k, v in o.items()}
        if isinstance(o, list):
            return [_walk(i) for i in o]
        return o
    return _walk(doc)

# Get all orders (admin only)
async def get_all_orders():
    try:
        coll = db.get_collection("orders")
        cursor = coll.find({}).sort("_id", -1)
        orders = [_serialize(o) for o in cursor]
        print(f"[order_controller] get_all_orders returned {len(orders)} orders")
        return orders
    except Exception as e:
        print(f"Error fetching orders: {str(e)}")
        import traceback
        traceback.print_exc()
        raise HTTPException(status_code=500, detail=f"Error fetching orders: {str(e)}")

# Get user's orders
async def get_user_orders(user_id: str):
    try:
        coll = db.get_collection("orders")
        cursor = coll.find({"user": ObjectId(user_id)}).sort("createdAt", -1)
        orders = [_serialize(o) for o in cursor]
        return orders
    except Exception as e:
        print(f"Error fetching user orders: {str(e)}")
        raise HTTPException(status_code=500, detail=f"Error fetching user orders: {str(e)}")

# Get orders by guestId
async def get_orders_by_guest_id(guest_id: str):
    try:
        coll = db.get_collection("orders")
        cursor = coll.find({"guestId": guest_id}).sort("createdAt", -1)
        orders = [_serialize(o) for o in cursor]
        return orders
    except Exception as e:
        print(f"Error fetching guest orders: {str(e)}")
        raise HTTPException(status_code=500, detail=f"Error fetching guest orders: {str(e)}")

# Get single order by ID
async def get_order_by_id(order_id: str):
    try:
        coll = db.get_collection("orders")
        # Try by orderId first (guest order), then by _id
        order = coll.find_one({"orderId": order_id})
        if not order and ObjectId.is_valid(order_id):
            order = coll.find_one({"_id": ObjectId(order_id)})
        if not order:
            raise HTTPException(status_code=404, detail="Order not found")
        return _serialize(order)
    except Exception as e:
        print(f"Error fetching order: {str(e)}")
        raise HTTPException(status_code=500, detail=f"Error fetching order: {str(e)}")

# Create new order
async def create_order(data: dict, user=None):
    try:
        coll = db.get_collection("orders")
        order_doc = {
            "orderId": data.get("orderId", f"ORD-{ObjectId()}"),
            "user": ObjectId(user.get("id")) if user and user.get("id") else None,
            "guestId": data.get("guestId"),
            "orderItems": data.get("orderItems", []),
            "contactDetails": data.get("contactDetails", {}),
            "shippingAddress": data.get("shippingAddress", {}),
            "billingAddress": data.get("billingAddress"),
            "paymentMethod": data.get("paymentMethod"),
            "itemsPrice": data.get("itemsPrice", 0),
            "taxPrice": data.get("taxPrice", 0),
            "shippingPrice": data.get("shippingPrice", 0),
            "totalPrice": data.get("totalPrice", data.get("totalAmount", 0)),
            "status": "pending",
            "createdAt": datetime.now(),
        }
        result = coll.insert_one(order_doc)
        order_doc["_id"] = str(result.inserted_id)
        return _serialize(order_doc)
    except Exception as e:
        print(f"Error creating order: {str(e)}")
        raise HTTPException(status_code=500, detail=f"Error creating order: {str(e)}")

# Update order status
async def update_order_status(order_id: str, data: dict, user=None):
    try:
        coll = db.get_collection("orders")
        order_oid = ObjectId(order_id) if ObjectId.is_valid(order_id) else None
        query = {"_id": order_oid} if order_oid else {"orderId": order_id}
        update = {"$set": {"status": data.get("status"), "updatedAt": datetime.now()}}
        result = coll.find_one_and_update(query, update, return_document=True)
        if not result:
            raise HTTPException(status_code=404, detail="Order not found")
        return _serialize(result)
    except Exception as e:
        print(f"Error updating order status: {str(e)}")
        raise HTTPException(status_code=500, detail=f"Error updating order status: {str(e)}")

# Update fulfillment status
async def update_fulfillment_status(order_id: str, data: dict, user=None):
    try:
        coll = db.get_collection("orders")
        order_oid = ObjectId(order_id) if ObjectId.is_valid(order_id) else None
        query = {"_id": order_oid} if order_oid else {"orderId": order_id}
        update = {"$set": {"fulfillmentStatus": data.get("fulfillmentStatus"), "updatedAt": datetime.now()}}
        result = coll.find_one_and_update(query, update, return_document=True)
        if not result:
            raise HTTPException(status_code=404, detail="Order not found")
        return _serialize(result)
    except Exception as e:
        print(f"Error updating fulfillment status: {str(e)}")
        raise HTTPException(status_code=500, detail=f"Error updating fulfillment status: {str(e)}")

# Update delivery status
async def update_delivery_status(order_id: str, data: dict, user=None):
    try:
        coll = db.get_collection("orders")
        order_oid = ObjectId(order_id) if ObjectId.is_valid(order_id) else None
        query = {"_id": order_oid} if order_oid else {"orderId": order_id}
        update = {"$set": {"deliveryStatus": data.get("deliveryStatus"), "updatedAt": datetime.now()}}
        result = coll.find_one_and_update(query, update, return_document=True)
        if not result:
            raise HTTPException(status_code=404, detail="Order not found")
        return _serialize(result)
    except Exception as e:
        print(f"Error updating delivery status: {str(e)}")
        raise HTTPException(status_code=500, detail=f"Error updating delivery status: {str(e)}")

# Update payment status
async def update_payment_status(order_id: str, data: dict, user=None):
    try:
        coll = db.get_collection("orders")
        order_oid = ObjectId(order_id) if ObjectId.is_valid(order_id) else None
        query = {"_id": order_oid} if order_oid else {"orderId": order_id}
        is_paid = data.get("isPaid", False)
        paid_at = datetime.fromtimestamp(data.get("paidAt") / 1000) if data.get("paidAt") else None
        update = {"$set": {"isPaid": is_paid, "paidAt": paid_at, "updatedAt": datetime.now()}}
        result = coll.find_one_and_update(query, update, return_document=True)
        if not result:
            raise HTTPException(status_code=404, detail="Order not found")
        return _serialize(result)
    except Exception as e:
        print(f"Error updating payment status: {str(e)}")
        raise HTTPException(status_code=500, detail=f"Error updating payment status: {str(e)}")

# Delete order
async def delete_order(order_id: str, user=None):
    try:
        coll = db.get_collection("orders")
        order_oid = ObjectId(order_id) if ObjectId.is_valid(order_id) else None
        query = {"_id": order_oid} if order_oid else {"orderId": order_id}
        result = coll.find_one_and_delete(query)
        if not result:
            raise HTTPException(status_code=404, detail="Order not found")
        return {"message": "Order deleted successfully"}
    except Exception as e:
        print(f"Error deleting order: {str(e)}")
        raise HTTPException(status_code=500, detail=f"Error deleting order: {str(e)}")

File: server/controllers/orderController.js
const Order = require('../models/Order');
const crypto = require('crypto');
const { sendOrderStatusUpdateEmail, sendOrderWithPDF } = require('../utils/emailService');
const { generateOrderPDF } = require('../utils/pdfGenerator');
const EventLog = require('../models/EventLog');

// Get all orders (admin only)
const getAllOrders = async (req, res) => {
  try {
    const orders = await Order.find()
      .populate('user', 'name email')
      .populate('orderItems.product', 'name price image')
      .sort({ createdAt: -1 });
    res.json(orders);
  } catch (error) {
    res.status(500).json({ message: error.message });
  }
};

// Get user's orders
const getUserOrders = async (req, res) => {
  try {
    const orders = await Order.find({ user: req.user.id })
      .populate('orderItems.product', 'name price image')
      .sort({ createdAt: -1 });
    res.json(orders);
  } catch (error) {
    res.status(500).json({ message: error.message });
  }
};

// Get orders by guestId (for order history without login)
const getOrdersByGuestId = async (req, res) => {
  try {
    const guestId = req.guestId;

    const orders = await Order.find({ guestId })
      .populate('orderItems.product', 'name price image')
      .sort({ createdAt: -1 });

    res.json(orders);
  } catch (error) {
    res.status(500).json({ message: error.message });
  }
};

// Get single order (anyone can view any order)
const getOrderById = async (req, res) => {
  try {
    const { id } = req.params;
    let order = null;

    // First try to find by orderId (guest order lookup)
    if (id && typeof id === 'string' && id.startsWith('ORD-')) {
      order = await Order.findOne({ orderId: id })
        .populate('user', 'name email')
        .populate('orderItems.product', 'name price image');
    }

    // If not found and id looks like MongoDB ObjectId, try findById
    if (!order && id && id.match(/^[0-9a-fA-F]{24}$/)) {
      order = await Order.findById(id)
        .populate('user', 'name email')
        .populate('orderItems.product', 'name price image');
    }

    if (!order) {
      return res.status(404).json({ message: 'Order not found' });
    }

    res.json(order);
  } catch (error) {
    res.status(500).json({ message: error.message });
  }
};

// Create new order (guest or authenticated users)
const createOrder = async (req, res) => {
  const { orderItems, shippingAddress, paymentMethod, itemsPrice, taxPrice, shippingPrice, totalAmount, billingAddress, couponCode, discountAmount } = req.body;

  try {
    const guestId = req.guestId;
    const userId = req.user?._id;

    if (!orderItems || orderItems.length === 0) {
      return res.status(400).json({ message: 'No order items' });
    }

    // create deterministic-ish unique order id
    const orderId = `ORD-${Date.now().toString(36)}-${crypto.randomBytes(4).toString('hex')}`;

    // Clean selectedVariants for non-variant products before saving
    const cleanedOrderItems = await Promise.all(orderItems.map(async (it) => {
      const itemData = {
        name: it.name,
        qty: it.qty || it.quantity || 1,
        price: it.price,
        product: it.product || it._id || null,
        image: it.image,
        variantImage: it.variantImage || null,
        selectedVariants: it.selectedVariants || null,
        selectedSize: it.selectedSize || null,
        selectedColor: it.selectedColor || null,
        colorCode: it.colorCode || null,
        variant: it.variant || null,
        sku: it.sku || null,
        variantId: it.variantId || null,
      };

      // Check if this product actually has variants defined
      if (itemData.product) {
        try {
          const Product = (await import('../models/Product.js')).default;
          const product = await Product.findById(itemData.product).select('variants');

          // If product has no variants array or empty variants, clear selectedVariants
          if (!product || !product.variants || product.variants.length === 0) {
            itemData.selectedVariants = null;
            itemData.variant = null;
          }
        } catch (err) {

          // On error, be conservative and keep the data as-is
        }
      }

      return itemData;
    }));

    const order = new Order({
      orderId,
      user: userId || null, // Allow null user for guest orders
      guestId: guestId, // Always set guestId for tracking
      orderItems: cleanedOrderItems,
      contactDetails: {
        firstName: shippingAddress?.firstName || '',
        lastName: shippingAddress?.lastName || '',
        email: shippingAddress?.email || '',
        phone: shippingAddress?.phone || '',
      },
      shippingAddress: {
        address: shippingAddress?.address || '',
        apartment: shippingAddress?.apartment || '',
        city: shippingAddress?.city || '',
        stateRegion: shippingAddress?.stateRegion || '',
        postalCode: shippingAddress?.postalCode || '',
        country: shippingAddress?.country || '',
      },
      billingAddress: billingAddress ? {
        address: billingAddress?.address || '',
        apartment: billingAddress?.apartment || '',
        city: billingAddress?.city || '',
        stateRegion: billingAddress?.stateRegion || '',
        postalCode: billingAddress?.postalCode || '',
        country: billingAddress?.country || '',
      } : null,
      paymentMethod,
      itemsPrice,
      taxPrice,
      shippingPrice,
      totalPrice: totalAmount,
      couponCode: couponCode || null,
      discountAmount: discountAmount || 0,
      status: 'pending',
    });

    const createdOrder = await order.save();

    // Log purchase event
    try {
      await EventLog.create({
        guestId,
        eventType: 'purchase',
        metadata: {
          orderId: createdOrder.orderId,
          cartValue: totalAmount,
          itemCount: orderItems.length,
        },
      });
    } catch (eventErr) {

    }

    res.status(201).json(createdOrder);
  } catch (error) {
    res.status(500).json({ message: error.message });
  }
};

// Update order status (admin only)
const updateOrderStatus = async (req, res) => {
  try {
    const { status } = req.body;

    if (!['pending', 'processing', 'shipped', 'completed', 'cancelled'].includes(status)) {
      return res.status(400).json({ message: 'Invalid status' });
    }

    const order = await Order.findByIdAndUpdate(
      req.params.id,
      { status, updatedAt: Date.now() },
      { new: true }
    );

    if (!order) {
      return res.status(404).json({ message: 'Order not found' });
    }

    // Send status update email to customer (in background - don't wait)
    sendOrderStatusUpdateEmail(order, status).catch(emailErr => {
      console.error(`✗ Failed to send status update email for order ${order.orderId}:`, emailErr.message);
    });

    res.json(order);
  } catch (error) {
    res.status(500).json({ message: error.message });
  }
};

// Update order payment status (admin only)
const updateOrderPayment = async (req, res) => {
  try {
    const { isPaid, paidAt } = req.body;

    const order = await Order.findByIdAndUpdate(
      req.params.id,
      { isPaid, paidAt: isPaid ? paidAt || Date.now() : null },
      { new: true }
    );

    if (!order) {
      return res.status(404).json({ message: 'Order not found' });
    }

    res.json(order);
  } catch (error) {
    res.status(500).json({ message: error.message });
  }
};

// Update order delivery status (admin only)
const updateOrderDelivery = async (req, res) => {
  try {
    const { deliveryStatus } = req.body;

    const updateData = { deliveryStatus };

    // Set deliveredAt timestamp when marking as delivered
    if (deliveryStatus === 'delivered') {
      updateData.deliveredAt = Date.now();
    }

    const order = await Order.findByIdAndUpdate(
      req.params.id,
      updateData,
      { new: true }
    );

    if (!order) {
      return res.status(404).json({ message: 'Order not found' });
    }

    // Send status update email to customer (in background - don't wait)
    sendOrderStatusUpdateEmail(order).catch(emailErr => {
      console.error(`✗ Failed to send delivery status email for order ${order.orderId}:`, emailErr.message);
    });

    res.json(order);
  } catch (error) {
    res.status(500).json({ message: error.message });
  }
};

// Update order refund status (admin only)
const updateOrderRefund = async (req, res) => {
  try {
    const { deliveryStatus } = req.body;

    const updateData = { deliveryStatus };

    const order = await Order.findByIdAndUpdate(
      req.params.id,
      updateData,
      { new: true }
    );

    if (!order) {
      return res.status(404).json({ message: 'Order not found' });
    }

    // Send status update email to customer (in background - don't wait)
    sendOrderStatusUpdateEmail(order).catch(emailErr => {
      console.error(`✗ Failed to send refund status email for order ${order.orderId}:`, emailErr.message);
    });

    res.json(order);
  } catch (error) {
    res.status(500).json({ message: error.message });
  }
};

// Update order fulfillment status (admin only)
const updateOrderFulfillment = async (req, res) => {
  try {
    const { fulfillmentStatus } = req.body;

    if (!['unfulfilled', 'fulfilled'].includes(fulfillmentStatus)) {
      return res.status(400).json({ message: 'Invalid fulfillment status' });
    }

    const order = await Order.findByIdAndUpdate(
      req.params.id,
      { fulfillmentStatus },
      { new: true }
    );

    if (!order) {
      return res.status(404).json({ message: 'Order not found' });
    }

    // Send status update email to customer (in background - don't wait)
    sendOrderStatusUpdateEmail(order).catch(emailErr => {
      console.error(`✗ Failed to send fulfillment status email for order ${order.orderId}:`, emailErr.message);
    });

    res.json(order);
  } catch (error) {
    res.status(500).json({ message: error.message });
  }
};

// Delete order (admin only)
const deleteOrder = async (req, res) => {
  try {
    const order = await Order.findByIdAndDelete(req.params.id);

    if (!order) {
      return res.status(404).json({ message: 'Order not found' });
    }

    res.json({ message: 'Order deleted successfully' });
  } catch (error) {
    res.status(500).json({ message: error.message });
  }
};

// Update order remarks (admin only)
const updateOrderRemarks = async (req, res) => {
  try {
    const { remarks } = req.body;

    const order = await Order.findByIdAndUpdate(
      req.params.id,
      { remarks: remarks || '' },
      { new: true }
    );

    if (!order) {
      return res.status(404).json({ message: 'Order not found' });
    }

    res.json(order);
  } catch (error) {
    res.status(500).json({ message: error.message });
  }
};

// Update contact details (admin only)
const updateOrderContact = async (req, res) => {
  try {
    const { contactDetails } = req.body;

    const order = await Order.findByIdAndUpdate(
      req.params.id,
      { contactDetails },
      { new: true }
    );

    if (!order) {
      return res.status(404).json({ message: 'Order not found' });
    }

    res.json(order);
  } catch (error) {
    res.status(500).json({ message: error.message });
  }
};

// Update shipping address (admin only)
const updateOrderShipping = async (req, res) => {
  try {
    const { shippingAddress } = req.body;

    const order = await Order.findByIdAndUpdate(
      req.params.id,
      { shippingAddress },
      { new: true }
    );

    if (!order) {
      return res.status(404).json({ message: 'Order not found' });
    }

    res.json(order);
  } catch (error) {
    res.status(500).json({ message: error.message });
  }
};

// Update billing address (admin only)
const updateOrderBilling = async (req, res) => {
  try {
    const { billingAddress } = req.body;

    const order = await Order.findByIdAndUpdate(
      req.params.id,
      { billingAddress },
      { new: true }
    );

    if (!order) {
      return res.status(404).json({ message: 'Order not found' });
    }

    res.json(order);
  } catch (error) {
    res.status(500).json({ message: error.message });
  }
};

// Resend order invoice PDF to customer (admin only)
const resendOrderPDF = async (req, res) => {
  try {
    const order = await Order.findById(req.params.id);

    if (!order) {
      return res.status(404).json({ message: 'Order not found' });
    }

    // Generate PDF and send
    const pdfBuffer = await generateOrderPDF(order);
    await sendOrderWithPDF(order, pdfBuffer);

    res.json({ message: 'Order invoice PDF sent successfully to customer' });
  } catch (error) {
    res.status(500).json({ message: error.message });
  }
};

module.exports = {
  getAllOrders,
  getUserOrders,
  getOrdersByGuestId,
  getOrderById,
  createOrder,
  updateOrderStatus,
  updateOrderPayment,
  updateOrderDelivery,
  updateOrderRefund,
  updateOrderFulfillment,
  deleteOrder,
  updateOrderRemarks,
  updateOrderContact,
  updateOrderShipping,
  updateOrderBilling,
  resendOrderPDF
};

